src/hotspot/share/runtime/vmThread.cpp

```diff
@@ -403,20 +404,28 @@ void VMThread::inner_execute(VM_Operation* op) {
                       _cur_vm_operation->name());
 
   bool end_safepoint = false;
+  bool should_abort_on_timeout = AbortVMOnVMOperationTimeout;
+  jlong vm_op_start_ns;
   if (_cur_vm_operation->evaluate_at_safepoint() &&
       !SafepointSynchronize::is_at_safepoint()) {
     SafepointSynchronize::begin();
-    if (_timeout_task != NULL) {
-      _timeout_task->arm();
+    if (should_abort_on_timeout) {
+      assert(_timeout_task != nullptr, "must created");
+      vm_op_start_ns = _timeout_task->arm();
     }
     end_safepoint = true;
   }
 
   evaluate_operation(_cur_vm_operation);
 
   if (end_safepoint) {
-    if (_timeout_task != NULL) {
+    if (should_abort_on_timeout) {
       _timeout_task->disarm();
+      jlong delay = nanos_to_millis(os::javaTimeNanos() - vm_op_start_ns);
+      if (delay > AbortVMOnVMOperationTimeoutDelay) {
+        fatal("%s VM operation took too long: completed in " JLONG_FORMAT " ms (timeout: " INTX_FORMAT " ms)",
+              _cur_vm_operation->name(), delay, AbortVMOnVMOperationTimeoutDelay);
+      }
```

Comment:
This duplicates `VMOperationTimeoutTask::task` body. So I think this is better to be moved to a separate method, e.g. `VMOperationTimeoutTask::check_and_fail`, which we can then call from both `::task` and `::disarm`.

Reply:
After this, we would also not need any other changes here: it would allow to check that task is `NULL` all the same.

Reply:
Note the fatal msg is different: on VM thread we know the VM-op has completed, but on TimeoutTask the VM-op status is unknown.

How about sth like this?

```cpp
bool check_timeout(jlong& delay) {
  delay = nanos_to_millis(os::javaTimeNanos() - _arm_time);
  return delay > AbortVMOnVMOperationTimeoutDelay
}

void disarm() {
  Atomic::release_store_fence(&_armed, 0);
  
  jlong delay;
  if (check_timeout(delay)) {
    fatal("%s VM operation took too long: completed in " JLONG_FORMAT " ms (timeout: " INTX_FORMAT " ms)",
        _cur_vm_operation->name(), delay, AbortVMOnVMOperationTimeoutDelay);
  }
}
```

Reply:
Huh, I missed this difference. Yeah, that goes in the right direction. Can we just pass the `const char*` with the format string to checking method? So we don't do the out-parameter, and the code ends up being:

```
bool maybe_fail_with(const char* fail_msg) {
  jlong delay = nanos_to_millis(os::javaTimeNanos() - _arm_time);
  if (delay > AbortVMOnVMOperationTimeoutDelay) {
     fatal(fail_msg, delay, AbortVMOnVMOperationTimeoutDelay);
  }
}

void disarm() {
  Atomic::release_store_fence(&_armed, 0);
  maybe_fail_with("VM operation took too long: completed in " JLONG_FORMAT " ms (timeout: " INTX_FORMAT " ms)");
}
```

Reply:
I actually think having the format string and the variables in the same scope/statement is better, because one can match the format specifier with the corresponding variable easily.

Reply:
I have no concerns about "duplicate" related checks here - it just isn't worth overthinking/over-engineering this IMO.